ntp-daemon/src/spawn/mod.rs

```diff
@@ -0,0 +1,240 @@
+pub enum PeerRemovalReason {
+    Demobilized,
+    NetworkIssue,
+}
+
+impl SpawnEvent {
+    pub fn new(id: SpawnerId, action: SpawnAction) -> SpawnEvent {
+        SpawnEvent { id, action }
+    }
+}
+
+/// The kind of action that the spawner requests to the system
+/// Currently a spawner can only create peers
+#[derive(Debug)]
+pub enum SpawnAction {
+    Create(PeerCreateParameters),
+    // Remove(()),
+}
+
+impl SpawnAction {
+    pub fn create(
+        id: PeerId,
+        addr: SocketAddr,
+        normalized_addr: NormalizedAddress,
+        nts: Option<PeerNtsData>,
+    ) -> SpawnAction {
+        SpawnAction::Create(PeerCreateParameters {
+            id,
+            addr,
+            normalized_addr,
+            nts,
+        })
+    }
+}
+
+#[derive(Debug)]
+pub struct PeerCreateParameters {
+    pub id: PeerId,
+    pub addr: SocketAddr,
+    pub normalized_addr: NormalizedAddress,
+    pub nts: Option<PeerNtsData>,
+}
+
+#[cfg(test)]
+impl PeerCreateParameters {
+    pub fn from_addr(addr: SocketAddr) -> PeerCreateParameters {
+        PeerCreateParameters {
+            id: PeerId::new(),
+            addr,
+            normalized_addr: NormalizedAddress::from_string_ntp(format!(
+                "{}:{}",
+                addr.ip(),
+                addr.port()
+            ))
+            .unwrap(),
+            nts: None,
+        }
+    }
+
+    pub fn from_ip_and_port(ip: impl Into<String>, port: u16) -> PeerCreateParameters {
+        Self::from_addr(SocketAddr::new(
+            ip.into().parse().expect("Invalid ip address specified"),
+            port,
+        ))
+    }
+
+    pub async fn from_normalized(addr: NormalizedAddress) -> PeerCreateParameters {
+        let socket_addr = addr
+            .lookup_host()
+            .await
+            .expect("Lookup failed")
+            .next()
+            .expect("Lookup unexpectedly returned zero responses");
+        PeerCreateParameters {
+            id: PeerId::new(),
+            addr: socket_addr,
+            normalized_addr: addr,
+            nts: None,
+        }
+    }
+}
+
+pub trait Spawner {
+    fn run(self, action_tx: mpsc::Sender<SpawnEvent>, system_notify: mpsc::Receiver<SystemEvent>);
+    fn get_id(&self) -> SpawnerId;
+    fn get_addr_description(&self) -> String;
+}
+
+#[async_trait::async_trait]
+pub trait BasicSpawner {
+    type Error: std::error::Error;
+    async fn handle_init(
+        &mut self,
+        action_tx: &mpsc::Sender<SpawnEvent>,
+    ) -> Result<(), Self::Error>;
+    async fn handle_peer_removed(
+        &mut self,
+        event: PeerRemovedEvent,
+        action_tx: &mpsc::Sender<SpawnEvent>,
+    ) -> Result<(), Self::Error>;
+    async fn handle_registered(
+        &mut self,
+        _event: PeerCreateParameters,
+        _action_tx: &mpsc::Sender<SpawnEvent>,
+    ) -> Result<(), Self::Error> {
+        Ok(())
+    }
+
+    fn get_id(&self) -> SpawnerId;
+    fn get_addr_description(&self) -> String;
+}
```

Comment:
trait functions should have some documentation on the expected contract and invariants